system/masking/include/datamaskingengine.hpp

```diff
@@ -0,0 +1,611 @@
+            return (profile ? profile->getUsedProperties(context) : emptyText.getLink());
+        }
+        virtual bool hasRule(const char* contentType, const IDataMaskingProfileContext* context) const override
+        {
+            IDataMaskingProfile* profile = selectProfile(context);
+            return (profile ? profile->hasRule(contentType, context) : false);
+        }
+        virtual size_t countOfRules(const char* contentType, const IDataMaskingProfileContext* context) const override
+        {
+            IDataMaskingProfile* profile = selectProfile(context);
+            return (profile ? profile->countOfRules(contentType, context) : 0);
+        }
+        virtual bool maskValue(const char* valueType, const char* maskStyle, char* buffer, size_t offset, size_t length, const IDataMaskingProfileContext* context, bool conditionally) const override
+        {
+            IDataMaskingProfile* profile = selectProfile(context);
+            return (profile ? profile->maskValue(valueType, maskStyle, buffer, offset, length, context, conditionally) : false);
+        }
+        virtual bool maskContent(const char* contentType, char* buffer, size_t offset, size_t length, const IDataMaskingProfileContext* context) const override
+        {
+            IDataMaskingProfile* profile = selectProfile(context);
+            return (profile ? profile->maskContent(contentType, buffer, offset, length, context) : false);
+        }
+        virtual bool getMarkupValueInfo(const char* element, const char* attribute, const char* buffer, size_t offset, size_t length, const IDataMaskingProfileContext* context, IDataMaskingDependencyCallback& callback, DataMaskingMarkupValueInfo& info) const
+        {
+            IDataMaskingProfile* profile = selectProfile(context);
+            return (profile ? profile->getMarkupValueInfo(element, attribute, buffer, offset, length, context, callback, info) : false);
+        }
+        virtual bool maskMarkupValue(const char* element, const char* attribute, char* buffer, size_t offset, size_t length, const IDataMaskingProfileContext* context, IDataMaskingDependencyCallback& callback) const override
+        {
+            IDataMaskingProfile* profile = selectProfile(context);
+            return (profile ? profile->maskMarkupValue(element, attribute, buffer, offset, length, context, callback) : false);
+        }
+        
+        virtual size_t loadProfiles(const IPTree& configuration) override
+        {
+            bool failed = false;
+            size_t addedProfiles = 0;
+            size_t addedDomains = 0;
+            if (streq(configuration.queryName(), "MaskingPlugin"))
+            {
+                if (!load(configuration, addedDomains, addedProfiles))
+                    failed = true;
+            }
+            else
+            {
+                Owned<IPTreeIterator> pluginConfigs(configuration.getElements("//MaskingPlugin"));
```

Comment:
Can MaskingPlugin really exist anywhere in the tree?

Reply:
As written, it apparently can. I must have been feeling generous when I wrote it, allowing some leeway in acceptable document structure. I might have been thinking a file which could be used with the manifest tool for inclusion in a binding might wrap the elements.